File: sdk/dotnet/Ancora.Tests/Examples/SharedHelpers.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;
using Ancora;

namespace Ancora.Tests.Examples;

/// <summary>
/// Shared helper types used across offline example tests.
/// No native library required -- the offline runtime produces events in-process.
/// </summary>

/// <summary>
/// An in-process run journal that records run IDs and event payloads.
/// Mirrors the role of a SQLite or Redis store in production.
/// </summary>
public sealed class RunJournal
{
    private readonly Dictionary<string, List<string>> _runs = new();

    public void RecordRun(string runId)
    {
        if (!_runs.ContainsKey(runId))
            _runs[runId] = new List<string>();
    }

    public void AppendEvent(string runId, string payload)
    {
        if (_runs.TryGetValue(runId, out var list))
            list.Add(payload);
    }

    public IReadOnlyList<string> EventsForRun(string runId)
        => _runs.TryGetValue(runId, out var list) ? list : Array.Empty<string>();

    public int RunCount => _runs.Count;
}

/// <summary>
/// Minimal stand-in for an OTEL span, to avoid importing OTEL packages.
/// </summary>
public sealed class Span
{
    private readonly long _startedAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
    public string Name { get; }
    public Dictionary<string, object> Attributes { get; } = new();

    public Span(string name) => Name = name;

    public void SetAttribute(string key, object value) => Attributes[key] = value;

    public long EndMs() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - _startedAt;
}

/// <summary>
/// Token estimation heuristic: 4 characters per token.
/// </summary>
public static class TokenEstimator
{
    public static int EstimateTokens(string text)
        => Math.Max(1, (int)Math.Ceiling(text.Length / 4.0));
}

File: sdk/java/src/test/java/io/ancora/examples/SharedHelpers.java
package io.ancora.examples;

import java.time.Instant;
import java.util.*;

/** Shared helpers used across the Java example tests. */
public final class SharedHelpers {

    private SharedHelpers() {}

    // ------------------------------------------------------------------
    // RunJournal
    // ------------------------------------------------------------------

    /** In-memory event store that mimics a durable restart journal. */
    public static final class RunJournal {
        private final Map<String, List<String>> store = new LinkedHashMap<>();

        /** Register a run ID. Idempotent: calling twice keeps count at 1. */
        public void recordRun(String runId) {
            store.putIfAbsent(runId, new ArrayList<>());
        }

        public void appendEvent(String runId, String eventJson) {
            store.computeIfAbsent(runId, k -> new ArrayList<>()).add(eventJson);
        }

        public List<String> eventsForRun(String runId) {
            return Collections.unmodifiableList(store.getOrDefault(runId, List.of()));
        }

        public int runCount() {
            return store.size();
        }
    }

    // ------------------------------------------------------------------
    // Span
    // ------------------------------------------------------------------

    /** Lightweight in-process span mirroring what an OTEL exporter would consume. */
    public static final class Span {
        private final String name;
        private final Instant start = Instant.now();
        private final Map<String, Object> attributes = new LinkedHashMap<>();
        private long durationMs = -1;

        public Span(String name) {
            this.name = name;
        }

        public String name() {
            return name;
        }

        public void setAttribute(String key, Object value) {
            attributes.put(key, value);
        }

        public Object getAttribute(String key) {
            return attributes.get(key);
        }

        public Map<String, Object> attributes() {
            return Collections.unmodifiableMap(attributes);
        }

        /** Marks the span as ended and returns elapsed milliseconds. */
        public long endMs() {
            durationMs = Instant.now().toEpochMilli() - start.toEpochMilli();
            return durationMs;
        }

        public long durationMs() {
            return durationMs;
        }
    }

    // ------------------------------------------------------------------
    // TokenEstimator
    // ------------------------------------------------------------------

    /** Rough 4-chars-per-token estimator for offline cost tracking. */
    public static final class TokenEstimator {
        private TokenEstimator() {}

        public static int estimateTokens(String text) {
            if (text == null || text.isEmpty()) return 1;
            return (int) Math.max(1, Math.ceil(text.length() / 4.0));
        }
    }
}
